src/internal/small_vec.rs

```diff
@@ -0,0 +1,87 @@
+use std::fmt;
+
+#[derive(Clone)]
+pub enum SmallVec<T> {
+    Empty,
+    One([T; 1]),
+    Two([T; 2]),
+    Flexible(Vec<T>),
+}
+
+impl<T> SmallVec<T> {
+    pub fn empty() -> Self {
+        Self::Empty
+    }
+
+    pub fn one(t: T) -> Self {
+        Self::One([t])
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        match self {
+            Self::Empty => &[],
+            Self::One(v) => v,
+            Self::Two(v) => v,
+            Self::Flexible(v) => v,
+        }
+    }
+
+    pub fn push(&mut self, new: T) {
+        let s = std::mem::replace(self, Self::empty());
+        *self = match s {
+            Self::Empty => Self::One([new]),
+            Self::One([v1]) => Self::Two([v1, new]),
+            Self::Two([v1, v2]) => Self::Flexible(vec![v1, v2, new]),
+            Self::Flexible(mut v) => {
+                v.push(new);
+                Self::Flexible(v)
+            }
+        }
+    }
+
+    pub fn with_capacity(len: usize) -> Self {
+        Self::Flexible(Vec::with_capacity(len))
```

Comment:
Should we match against the size in case `len` is less than the flexible breakpoint?

Reply:
I thought I did. Looks like @aleksator removed it, why have it allocate if the len is breakpoint?